### DISCUSSION TITLE
Help with a simple car model

### DISCUSSION THREAD

**Discussion by aalhamada** [OpenModelica/OpenModelica#10733] [Q&A]
Greetings, 

I am new to Modelica and I have been working on a project to create a digital twin for a simple car using Modelica. 

My model would simply consist of Battery, PID controller, motor, brakes and tires. 

However, I kept running into issues and was looking around for tips to fix up my model which made it a little more complex with torque, inertia, and feedback connectors. 

I am currently running into a few sets of errors I'm stuck with. 

Everytime I run the model, I get these errors. 

----------------

1] 23:59:45 Translation Warning
[VehicleInterfaces.Brakes.MinimalBrakes: 88:3-88:49]: [An inner declaration for outer component world could not be found and was automatically generated.](omeditmessagesbrowser:///VehicleInterfaces.Brakes.MinimalBrakes?lineNumber=88)

[2] 23:59:45 Translation Notification
[VehicleInterfaces.Brakes.MinimalBrakes: 88:3-88:49]: The diagnostics message for the missing inner is: No "world" component is defined. A default world
component with the default gravity field will be used
(g=9.81 in negative y-axis). If this is not desired,
drag Modelica.Mechanics.MultiBody.World into the top level of your model.

[3] 23:59:45 Translation Error
[model2: 35:3-36:65]: The type of variables motor.flange_motor and torque.tau
are inconsistent in connecting equations.

----------------

I'm not sure what to do with either of these issues to fix the problem. Yet, I know that # refers to the vehicle interface I got for Modelica from the import section. (I just found it there and imported it..didn't download anything external)

Here is how my model looks and would love any feedback or tips to improve it and hopefully manage to make it work

![model2](https://github.com/OpenModelica/OpenModelica/assets/134484599/378c3b11-910c-44aa-a9c3-9b7c06eeb58b)

Please note that most of these components were used because of my current research on how to create a car model. I could be totally wrong.

**Comment by casella** [comment]
> 1] 23:59:45 Translation Warning [VehicleInterfaces.Brakes.MinimalBrakes: 88:3-88:49]: An inner declaration for outer component world could not be found and was automatically generated.
> 
> [2] 23:59:45 Translation Notification [VehicleInterfaces.Brakes.MinimalBrakes: 88:3-88:49]: The diagnostics message for the missing inner is: No "world" component is defined. A default world component with the default gravity field will be used (g=9.81 in negative y-axis). If this is not desired, drag Modelica.Mechanics.MultiBody.World into the top level of your model.

Look at the example models: they all have world, road, and atmosphere inner models. These define global parameters and variables that are accessible by other components without the need of explict connections. Please drag them into your system models.

> [3] 23:59:45 Translation Error [model2: 35:3-36:65]: The type of variables motor.flange_motor and torque.tau are inconsistent in connecting equations.

Here you are trying to connect two ports that do not have the same type (e.g. a rotational 1D flange), which of course makes no sense. If you post your example model code here,  I could give you more feedback.

**Reply by aalhamada** [reply]
Hello again, 

This is the code of my model. It is generated by modelica. I didn't change anything. 

```mo
model model2
  Modelica.Electrical.Batteries.BatteryStacks.CellStack cellStack annotation(
    Placement(visible = true, transformation(origin = {-89, 87}, extent = {{-9, -7}, {9, 7}}, rotation = 90)));
  VehicleInterfaces.Brakes.MinimalBrakes minimalBrakes annotation(
    Placement(visible = true, transformation(origin = {-35, 27}, extent = {{5, -5}, {-5, 5}}, rotation = 0)));
  Modelica.Mechanics.MultiBody.Visualizers.VoluminousWheel voluminousWheel annotation(
    Placement(visible = true, transformation(origin = {-62, 22}, extent = {{-4, -4}, {4, 4}}, rotation = 180)));
  Modelica.Mechanics.MultiBody.Visualizers.VoluminousWheel voluminousWheel1 annotation(
    Placement(visible = true, transformation(origin = {-12, 22}, extent = {{-4, -4}, {4, 4}}, rotation = 0)));
  Modelica.Mechanics.MultiBody.Visualizers.VoluminousWheel voluminousWheel2 annotation(
    Placement(visible = true, transformation(origin = {-62, 32}, extent = {{-4, -4}, {4, 4}}, rotation = 180)));
  Modelica.Mechanics.MultiBody.Visualizers.VoluminousWheel voluminousWheel3 annotation(
    Placement(visible = true, transformation(origin = {-12, 32}, extent = {{-4, -4}, {4, 4}}, rotation = 0)));
  Modelica.Mechanics.MultiBody.Examples.Systems.RobotR3.Utilities.Motor motor annotation(
    Placement(visible = true, transformation(origin = {-39, 67}, extent = {{-7, -7}, {7, 7}}, rotation = -90)));
  Modelica.Blocks.Continuous.PID pid annotation(
    Placement(visible = true, transformation(origin = {-3, 85}, extent = {{-3, -3}, {3, 3}}, rotation = 270)));
  Modelica.Mechanics.Rotational.Sources.Torque torque(useSupport = true)  annotation(
    Placement(visible = true, transformation(origin = {-14, 70}, extent = {{-4, -4}, {4, 4}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Components.Inertia inertia(J = 1)  annotation(
    Placement(visible = true, transformation(origin = {-1, 69}, extent = {{-5, -5}, {5, 5}}, rotation = 0)));
  Modelica.Blocks.Math.Feedback feedback annotation(
    Placement(visible = true, transformation(origin = {40, 82}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.Step step annotation(
    Placement(visible = true, transformation(origin = {14, 82}, extent = {{-4, -4}, {4, 4}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sensors.SpeedSensor speedSensor annotation(
    Placement(visible = true, transformation(origin = {11, 59}, extent = {{-3, -3}, {3, 3}}, rotation = 0)));
  Modelica.Blocks.Continuous.SecondOrder secondOrder annotation(
    Placement(visible = true, transformation(origin = {25, 69}, extent = {{-3, -3}, {3, 3}}, rotation = 0)));
equation
  connect(cellStack.heatPort, motor.axisControlBus) annotation(
    Line(points = {{-82, 87}, {-75, 87}, {-75, 67}, {-46, 67}}, color = {191, 0, 0}));
  connect(motor.flange_motor, minimalBrakes.brakePedal) annotation(
    Line(points = {{-39, 60}, {-36, 60}, {-36, 32}, {-35, 32}}));
  connect(torque.flange, inertia.flange_a) annotation(
    Line(points = {{-10, 70}, {-10, 69}, {-6, 69}}));
  connect(motor.flange_motor, torque.tau) annotation(
    Line(points = {{-39, 60}, {-26, 60}, {-26, 70}, {-18, 70}}));
  connect(step.y, feedback.u1) annotation(
    Line(points = {{18, 82}, {32, 82}}, color = {0, 0, 127}));
  connect(inertia.flange_b, speedSensor.flange) annotation(
    Line(points = {{4, 70}, {4, 59}, {8, 59}}));
  connect(motor.flange_motor, speedSensor.flange) annotation(
    Line(points = {{-39, 60}, {-21, 60}, {-21, 54}, {8, 54}, {8, 59}}));
  connect(feedback.y, pid.u) annotation(
    Line(points = {{49, 82}, {50, 82}, {50, 92}, {-3, 92}, {-3, 89}}, color = {0, 0, 127}));
  connect(minimalBrakes.wheelHub_2, voluminousWheel3.frame_a) annotation(
    Line(points = {{-32, 32}, {-16, 32}}, color = {135, 135, 135}));
  connect(minimalBrakes.wheelHub_4, voluminousWheel2.frame_a) annotation(
    Line(points = {{-38, 32}, {-58, 32}}, color = {135, 135, 135}));
  connect(minimalBrakes.wheelHub_1, voluminousWheel1.frame_a) annotation(
    Line(points = {{-32, 22}, {-16, 22}}, color = {135, 135, 135}));
  connect(minimalBrakes.wheelHub_3, voluminousWheel.frame_a) annotation(
    Line(points = {{-38, 22}, {-58, 22}}, color = {135, 135, 135}));
  connect(cellStack.heatPort, pid.u) annotation(
    Line(points = {{-82, 88}, {-10, 88}, {-10, 92}, {-3, 92}, {-3, 89}}, color = {191, 0, 0}));
  connect(pid.y, motor.flange_motor) annotation(
    Line(points = {{-3, 82}, {-50, 82}, {-50, 60}, {-39, 60}}, color = {0, 0, 127}));
  connect(pid.y, motor.axisControlBus) annotation(
    Line(points = {{-2, 82}, {-4, 82}, {-4, 76}, {-52, 76}, {-52, 67}, {-46, 67}}, color = {0, 0, 127}));
  connect(pid.y, secondOrder.u) annotation(
    Line(points = {{-2, 82}, {8, 82}, {8, 69}, {21, 69}}, color = {0, 0, 127}));
  connect(secondOrder.y, feedback.u2) annotation(
    Line(points = {{28, 70}, {40, 70}, {40, 74}}, color = {0, 0, 127}));
  connect(speedSensor.w, feedback.u2) annotation(
    Line(points = {{14, 59}, {40, 59}, {40, 74}}, color = {0, 0, 127}));
  annotation(
    uses(VehicleInterfaces(version = "2.0.1"), Modelica(version = "4.0.0")));
end model2;
```

1# : I get that but how to connect them? Or where to put them in the model? Since, If I don't then what is the point of them being there? I know a beginner's question but this is my first model with Modelica so I'm trying to work with the minimum to create a model then go from there.

2# : That is correct. What I am trying to do is simply connect the tyres to the motor and of course, the flange has to be involved to control the slowing down of the tyres. As in the model, I mistakingly connected the motor directly to the flange. What I am trying to do is to actually connect it to something to provide the rotation 'power' for tyres. However, I think I am missing something (a component) to connect the motor to then to the tyres in order to make the model work. 

What would you suggest in order to make this work?

**Reply by casella** [reply]
> 1# : I get that but how to connect them? 

There's no need to connect them. Those models will be automatically declared as `inner` (thanks to an annotation that instructs the editor to do so). Then, all the other sub-models in the same model will be able to access them by declaring them as `outer`. See, e.g. [these slides](https://www.google.com/url?sa=i&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=0CAIQw7AJahcKEwjI4bPzv5D_AhUAAAAAHQAAAAAQBA&url=https%3A%2F%2Frmc.dlr.de%2Frm%2Fde%2Fstaff%2Fextcms%2Fimages%2Frmc%2Fusers%2Fzimm_di%2Flecture%2FLecture9aHandout(1).pdf&psig=AOvVaw3r0i-ZeiHELANS8cVLVxV5&ust=1685105116979823).

> Or where to put them in the model? 

Anywhere, it doesn't matter.

> Since, If I don't then what is the point of them being there?

They contain global parameters and variables that are meant to be accessible by all submodels.

> 2# : That is correct. What I am trying to do is simply connect the tyres to the motor and of course, the flange has to be involved to control the slowing down of the tyres. As in the model, I mistakingly connected the motor directly to the flange. What I am trying to do is to actually connect it to something to provide the rotation 'power' for tyres. However, I think I am missing something (a component) to connect the motor to then to the tyres in order to make the model work.

The problem is this statement:
```modelica
connect(motor.flange_motor, torque.tau)
```
You are connecting a physical flange to a signal source, which doesn't make sense. You can only connect pyhsical flanges to other physical flanges of the same type.

You can use a torque generator as prime mover, but then you need to connect some _signal_ source to its tau input connector. A ramp, a step, a table-based interpolation, whatever you find in Modelica.Blocks.Sources.

**Reply by aalhamada** [reply]
I see. Now, I understand better. Thank you very much. I have been pretty confused and didn't think about this clearly. 

I'm working on that now. 

Also, I'm currently trying to use the example for the rear-drive car in the vehicle interface library. I wanted to switch the engine with a pid controller but every time I do that, I get the 'inconsistent' equation. Despite that, I know the PID controller needs to have throttle and steering signals in order to work.

I also wanted to add a GPS to the model. Would that be possible? If so, where would I start as I don't think I found anything directly related to GPS in the common Modelica libraries. 

I know I'm asking too many questions but I do appreciate the help a lot. Thank you
